### PR TITLE
Enable pulling testcases from Clusterfuzz and add extra sanity checks to CI

### PR DESCRIPTION
**Pull request by everestmz**
cc: @google/clusterfuzz 
- Added more sanity checks
- Enabled pulling new testcases from clusterfuzz
- Enable logging to stackdriver
- Reset after testcases now deletes output dirs, checks out chromium src to master and runs gclient runhooks + sync

### CODE REVIEW

File: ci/continuous_integration/daemon/main.py [hunk 1]
```diff
@@ -78,16 +77,73 @@ def update_auth_header():
   os.chmod(AUTH_FILE_LOCATION, 0600)
 
 
+def get_version():
+  """Returns the version of the binary."""
+
+  proc = subprocess.Popen(build_command('supported_job_types'), stdout=subprocess.PIPE, stderr=subprocess.PIPE, shell=True)
```
Comment (tanin47) [inline ci/continuous_integration/daemon/main.py]: Use `subprocess.check_output`. It'll return string.

File: ci/continuous_integration/daemon/main.py [hunk 2]
```diff
@@ -78,16 +77,73 @@ def update_auth_header():
...
+  out, _ = proc.communicate()
+  return yaml.load(out)['Version']
+
+
+def get_supported_jobtypes():
+  """Returns a hash of supported job types."""
+
+  proc = subprocess.Popen(build_command('supported_job_types'), stdout=subprocess.PIPE, stderr=subprocess.STDOUT, shell=True)
```
Comment (tanin47) [inline ci/continuous_integration/daemon/main.py]: Use `subprocess.check_output`. It'll return string.

File: ci/continuous_integration/daemon/main.py [hunk 3]
```diff
@@ -78,16 +77,73 @@ def update_auth_header():
...
+    r = requests.post('https://clusterfuzz.com/v2/testcases/load',
+                      headers={'Authorization': auth_header},
+                      json={'page': page,
+                            'reproducible': 'yes'})
+    testcases = json.loads(r.text)['items']
+    testcases = [testcase['id'] for testcase in testcases
+                 if testcase['jobType'] in supported_jobtypes['chromium']]
+    if latest_testcase in testcases:
+      testcases = testcases[0:testcases.index(latest_testcase)]
+    page = random.randint(5, 35)
```
Comment (tanin47) [inline ci/continuous_integration/daemon/main.py]: Why will we jump page randomly?

File: ci/continuous_integration/daemon/main.py [hunk 4]
```diff
@@ -78,16 +77,73 @@ def update_auth_header():
...
+
+  testcases = None
+  page = 1
+  supported_jobtypes = get_supported_jobtypes()
+  while not testcases:
+    r = requests.post('https://clusterfuzz.com/v2/testcases/load',
+                      headers={'Authorization': auth_header},
+                      json={'page': page,
+                            'reproducible': 'yes'})
+    testcases = json.loads(r.text)['items']
```
Comment (tanin47) [inline ci/continuous_integration/daemon/main.py]: Use `r.json()`

File: ci/continuous_integration/daemon/main.py [hunk 5]
```diff
@@ -78,16 +77,73 @@ def update_auth_header():
...
+    if latest_testcase in testcases:
+      testcases = testcases[0:testcases.index(latest_testcase)]
+    page = random.randint(5, 35)
+  return testcases
+
+
+def reset_and_run_testcase(testcase_id, test_type):
+  """Resets the chromium repo and runs the testcase."""
+
+  if os.path.exists(CHROMIUM_OUT):
```
Comment (tanin47) [inline ci/continuous_integration/daemon/main.py]: I think we should make a method for `delete_if_exists`.

File: ci/continuous_integration/daemon/main.py [hunk 6]
```diff
@@ -78,16 +77,73 @@ def update_auth_header():
...
+    shutil.rmtree(CHROMIUM_OUT)
+  if os.path.exists(CLUSTERFUZZ_DIR):
+    shutil.rmtree(CLUSTERFUZZ_DIR)
+  subprocess.call('git checkout -f master', shell=True, cwd=CHROMIUM_SRC)
```
Comment (tanin47) [inline ci/continuous_integration/daemon/main.py]: This one fails silently. Either use `check_call` or refactor it to `common.execute`.

File: ci/continuous_integration/daemon/main.py [hunk 7]
```diff
@@ -78,16 +77,73 @@ def update_auth_header():
...
+  subprocess.call('export PATH=$PATH:%s && gclient sync' % DEPOT_TOOLS, shell=True, cwd=CHROMIUM_SRC)
```
Comment (tanin47) [inline ci/continuous_integration/daemon/main.py]: Well, ok, we might want to have `common.execute` that always inject `DEPOT_TOOLS`. And use the `environment` param instead.

File: ci/playbook.yml [hunk 8]
```diff
@@ -104,9 +103,10 @@
       with_items:
         - {'src': 'python_daemon/run', 'dest': '/etc/sv/python-daemon/run'}
         - {'src': 'python_daemon/log_run', 'dest': '/etc/sv/python-daemon/log/run'}
-        - {'src': 'continuous_integration/ci', 'dest': '/python-daemon'}
-        - {'src': '/auto/FuzzInfrastructure/clusterfuzz-tools/clusterfuzz-tools-ci-credentials.json', 'dest': '/python-daemon/service-account-credentials.json'}
+        - {'src': 'continuous_integration/daemon', 'dest': '/python-daemon'}
+        - {'src': 'credentials.json', 'dest': '/python-daemon/service-account-credentials.json'}
         - {'src': '/auto/FuzzInfrastructure/clusterfuzz-tools/goma', 'dest': '/home/clusterfuzz'}
+        - {'src': '/google/data/rw/teams/clusterfuzz-tools/releases/clusterfuzz-rc', 'dest': '/python-daemon/clusterfuzz'}
```
Comment (tanin47) [inline ci/playbook.yml]: Use `ro` instead of `rw`

File: ci/python_daemon/run [hunk 9]
```diff
@@ -1,3 +1,3 @@
 #!/bin/bash
 exec 2>&1
-exec chpst -u clusterfuzz python /python-daemon/ci/main.py
+exec chpst -u clusterfuzz python /python-daemon/daemon/main.py
```
Comment (tanin47) [inline ci/python_daemon/run]: I am hoping that we would package the daemon into pex. Let's do that in the next CL.

File: ci/continuous_integration/daemon/main.py (hunk 3 above)
Comment (everestmz) [inline ci/continuous_integration/daemon/main.py]: We just want to kill some time running other testcases until page 1 has more new testcases again - in the next CL I will implement the robust caching system we discussed in 1:1

File: ci/continuous_integration/daemon/main.py (hunk 6 above)
Comment (everestmz) [inline ci/continuous_integration/daemon/main.py]: Does this mean pulling common into shared?

Comment (tanin47) [inline ci/continuous_integration/daemon/main.py]: Let's not do that. It seems `common.execute` is fairly specific to the tool.

**Comment by everestmz** [comment]
Updated!

File: ci/continuous_integration/daemon/main.py (hunk 3 above)
Comment (tanin47) [inline ci/continuous_integration/daemon/main.py]: Why can't we do `page += 1`? That seems to give more priority to the newer testcases.

File: ci/continuous_integration/daemon/main.py [hunk 10]
```diff
@@ -78,16 +77,86 @@ def update_auth_header():
...
+  """Delete filename if the file exists."""
+
+  if os.path.exists(filename):
+    shutil.rmtree(filename)
+
+
+def call_with_depot_tools(command, cwd=CHROMIUM_SRC):
+  """Run command with depot_tools in the path."""
+  environment = os.environ.copy()
+  print(environment)
```
Comment (tanin47) [inline ci/continuous_integration/daemon/main.py]: Remove

**Pull request merged**